### PR TITLE
Remove exception catching for __init__ functions, add InitError exception type

### PR DESCRIPTION
**Pull request by jdlangs**
This PR is to continue the discussion from #12505 with a concrete implementation candidate. Sometime later this week I am interested to try running the package evaluator to see what level of breakage it causes due to people not bothering to fix warnings.

@JeffBezanson @StefanKarpinski @vtjnash

### CODE REVIEW

**Comment by stevengj** [comment]
+1.  I don't think that code in `__init__` should be semantically different (in the absence of precompilation) from the same code pasted at the end of your module.

**Comment by ivarne** [comment]
Marking this as 0.4.0 as this is a breaking change, that should get merged before feature freeze.

**Comment by StefanKarpinski** [comment]
Indeed, I think this needs to get resolved.

**Comment by ScottPJones** [comment]
:+1:

**Comment by ivarne** [comment]
If it should behave like the same code at the end of a module, we need to catch the exception and wrap it in a `LoadError` or a new `InitError` (to store a reference to the failing module, and different printing).

**Comment by ScottPJones** [comment]
@ivarne That sounds good.  The principle here is that errors should not be swept under the rug (printing something to the terminal just doesn't cut it when you have lots of non-interactive processes anyway, and can be dangerous if the output device is been redirected to something else)

**Comment by StefanKarpinski** [comment]
I'm still confused by why we would ever be ignoring errors in `__init__` in the first place. Can you spell that out, @vtjnash?

**Comment by vtjnash** [comment]
Jeff did the implementation of this. I think the basic assumption was that a failure in a `__init__` may be non-critical for the state of the rest of the system. And if you ran:
```
try
  import Foo
  nofoo = !isdefined(current_module(), :Foo) # necessary since failure to define Foo isn't an error
catch e
  nofoo = true
end
```
and Foo did stuff like have multiple submodules with `__init__` and loading other modules, you want all of the unrelated ones to get a chance to finish running.

As some food for thought, it may not be correct to propagate errors for code loaded from conditional modules (https://github.com/JuliaLang/julia/pull/6884) since that code is not run until after the matching module.

**Comment by stevengj** [comment]
@ivarne, I agree with wrapping it in a `LoadError` to make it like other exceptions thrown by module-body code.  @phobon, this means keeping the try-catch block, but putting something like the following in the catch block:
```
jl_rethrow_other(jl_new_struct(jl_loaderror_type, f->linfo->file->name, f->linfo->line, jl_exception_in_transit));
```

**Comment by phobon** [comment]
It would definitely be nice to have `InitError` so that "expected" problems were detectable. I can easily see someone trying to import a module, seeing if its initializer fails (e.g., because of a missing library) and proceed to do something else. Of course, this would also require allowing `import`/`using` within `try` (at global scope only) but for now you can do it with `try reload(:BadMod) catch end`.

I should have some time this evening to update the branch.

**Comment by phobon** [comment]
Updated to add a new `InitError` type. Note that exceptions from `__init__` were being caught and put into a `LoadError` already when it was coming from a file. This checks specifically for a thrown `InitError` and rethrows it in that case. Would appreciate feedback on the implementation there.

**Comment by stevengj** [comment]
Maybe add
```jl
@test_throws InitError include_string("module TestInitError\n__init__() = error(\"foo\")\nend")
```
to `test/core.jl`?

**Comment by stevengj** [comment]
Probably you should change `precompilableerror(ex::LoadError, c)` in `base/loading.c` to
```jl
precompilableerror(ex::Union{LoadError,InitError}, c) = precompilableerror(ex.error, c)
```
to cover the (unlikely) event that the user calls `__precompile(false)__` from their `__init__` function instead of the module body.

File: base/replutil.jl [hunk 1]
```diff
@@ -93,6 +93,13 @@ function showerror(io::IO, ex::LoadError, bt; backtrace=true)
 end
 showerror(io::IO, ex::LoadError) = showerror(io, ex, [])
 
+function showerror(io::IO, ex::InitError, bt; backtrace=true)
+    print(io, "InitError: ")
+    showerror(io, ex.error, bt, backtrace=backtrace)
+    print(io, "\nduring initialization of module $(ex.mod)")
```
Comment (jakebolewski) [inline base/replutil.jl]: I feel this should be:
```julia
showerror(io::IO, ex::InitError) = print(io, "InitError: during initialization of module $(ex.mod"))
```
the backtrace will already be appended

Comment (stevengj) [inline base/replutil.jl]: @jakebolewski, It seems like @phobon's code is correct; it should be analogous to `showerror(io::IO, ex::LoadError)`, and you need to show `ex.error` at minimum.

**Comment by vtjnash** [comment]
> to cover the (unlikely) event that the user calls __precompile(false)__ from their __init__ function instead of the module body.

the precompilation has already finished and saved the file by then, thus it seems like a bad place to be trying to assert `!precompiled`, unless there is some reason to conditionally permit usage of the precompiled module?

**Comment by stevengj** [comment]
@vtjnash, I agree that we normally shouldn't encourage people to call `__precompile__` in `__init__`.  I was just going by the principle that code in `__init__` should be functionally equivalent to code pasted at the end of the module.

**Comment by phobon** [comment]
@stevengj, currently if you define a module at the REPL, exceptions do not get wrapped in a `LoadError`. Would it be desirable to also not wrap `__init__` exceptions in that case? Or is it the current non-wrapping behavior that should be changed?

**Comment by stevengj** [comment]
@phobon, as I understand it, `LoadError` is only something that gets thrown by `include` or `include_string` or by things that call this; it has nothing to do with whether the included file defines a module or not.   I don't think we want to throw `LoadError`s within a single file, and so I think it is fine not to do so for things evaluated at the REPL.

On the other hand, `__init__` is specific to modules, and I think it makes sense to wrap `__init__` exceptions in `InitError`, no matter how the module is created.

**Comment by vtjnash** [comment]
i think you need a try/catch/rethrow at https://github.com/phobon/julia/blob/ddd476b1b7e5c3af0d34432109d37693604b11c6/src/toplevel.c#L208-L215 to clear the reinit list after an error to avoid leaving it in an invalid state

**Comment by phobon** [comment]
@vtjnash, I'm not completely sure I know what "clear the reinit list" entails. Do these lines do the job?
```
JL_TRY {
    size_t i, l=module_stack.len;
    for(i = stackidx; i < l; i++) {
        jl_module_load_time_initialize((jl_module_t*)module_stack.items[i]);
    }
    assert(module_stack.len == l);
    module_stack.len = stackidx;
}
JL_CATCH {
    arraylist_free(&module_stack);
    initialized = 0;
    jl_rethrow();
}
```
Is the same concern relevant for restoring compiled modules as well?

**Comment by vtjnash** [comment]
clear it by resetting the top pointer: `module_stack.len = stackidx;`

precompiled modules have their own array, so i don't think they should be a problem

File: base/loading.jl [hunk 2]
```diff
@@ -145,7 +145,7 @@ function show(io::IO, ex::PrecompilableError)
     end
 end
 precompilableerror(ex::PrecompilableError, c) = ex.isprecompilable == c
-precompilableerror(ex::LoadError, c) = precompilableerror(ex.error, c)
+precompilableerror(ex::Union{LoadError,InitError}, c) = precompilableerror(ex.error, c)
 precompilableerror(ex, c) = false
```
Comment (ivarne) [inline base/loading.jl]: Not sure if this is useful for other cases, but this change kind of begs for `abstract WrappedException <: Exception`

Comment (phobon) [inline base/loading.jl]: I had thought about it, but it did seem like a pretty small corner case. I suppose there's no harm in future-proofing though.

File: base/loading.jl [hunk 3]
```diff
@@ -145,7 +145,7 @@ function show(io::IO, ex::PrecompilableError)
     end
 end
 precompilableerror(ex::PrecompilableError, c) = ex.isprecompilable == c
-precompilableerror(ex::LoadError, c) = precompilableerror(ex.error, c)
+precompilableerror(ex::WrappedException c) = precompilableerror(ex.error, c)
```
Comment (stevengj) [inline base/loading.jl]: missed a comma here

Comment (phobon) [inline base/loading.jl]: oops, that's embarrassing.

**Comment by phobon** [comment]
Rebased, squashed, and some docs added.

**Comment by jakebolewski** [comment]
Bump.  This looks good to merge, any objections?

**Comment by ivarne** [comment]
It will be interesting to see the ratio of "real bugs"  vs "bad code" issues that will be opened in the fallout from this change. 

A mention in `NEWS.md` would seem appropriate, but unless @phobon finds time to update the PR before merging, it can easily be done later.

**Comment by phobon** [comment]
I did have some interest in running this branch on the package evaluator but started to hesitate digging into when I saw it would require some hacking. Is there some fork or branch where someone has already set up this capability?

And just saw the merge went through...

**Comment by phobon** [comment]
Glad to see this resolved, thanks for the feedback everyone!

**Comment by jakebolewski** [comment]
Thanks @phobon, great work.

**Comment by tkelman** [comment]
> it would require some hacking

Not much? It's a few clicks on the buildbots by anyone who has a login to make a custom generic binary from any branch or PR, then should be only one or two lines to change in PackageEvaluator to point it to a different download.

**Comment by jakebolewski** [comment]
It will be great when this can be done just by pinging a bot from Github.

**Comment by tkelman** [comment]
Webhooks aren't hard to write. Just a matter of hardware to run it on and plumbing it together.

**Comment by jakebolewski** [comment]
Hopefully soon (end of the month) we will have the hardware.

**Comment by phobon** [comment]
It would be nice to enable PackageEvaluator to check out and build an arbitrary branch so anyone without buildbot access could run it locally, or maybe on their own cloud computing account.

**Comment by tkelman** [comment]
Recent incarnations of PackageEvaluator don't build from source any more, but the buildbots are good at that. You can run `make binary-dist` from anywhere and the product should run fine on an equivalent distro, but to make generic binaries that run across many different distributions requires some work and custom setup.

**Comment by phobon** [comment]
That's fine, but will they be freely usable by anyone who wants? If not, then in the interest of simplicity it would seem useful to provide an option to PackageEvaluator that would have it also build julia from source.

**Comment by IainNZ** [comment]
@phobon I'd accept a pull request for that

**Comment by tkelman** [comment]
> will they be freely usable by anyone who wants?

Custom builds from the buildbots get uploaded to https://s3.amazonaws.com/julianightlies/ with a sha, just like normal nightlies. They'll stay up there for roughly 3-4 weeks. In fact the releases of 0.3 have been made that way, just manually moving the files over from julianightlies to julialang once a month.

**Comment by tkelman** [comment]
Note that this change is really bad for embedding. Now when any dependency libraries aren't present, Julia can't start at all instead of being just unable to use the functionality from that library.

**Comment by ivarne** [comment]
In that case the library should `warn("Missing X dependency ") ` instead of `error("Missing X dependency ") `

**Comment by phobon** [comment]
Or it should be possible to put the `import` or `using` call in a try block.

**Comment by ivarne** [comment]
If we actually manage to agree on something to move out of Base in https://github.com/JuliaLang/julia/issues/5155, we wouldn't need to `__init__()` so many unused modules at startup time.

**Comment by tkelman** [comment]
PCRE, GMP, dSFMT, openblas, and suitesparse_wrapper all need to be turned into warnings instead of `InitError` when they are not present, to get a REPL to start.

**Comment by vtjnash** [comment]
Or pull them out of base? dSFMT adds significantly to Julia's startup time (also openblas, but less so)

**Comment by tkelman** [comment]
Yes, but not right now. We need to make the lack of their presence non-fatal right now, otherwise 0.4.0 will be substantially less usable for embedders.

**Comment by ScottPJones** [comment]
Hopefully pulling them out of base will be done early in 0.5, but this change is already in 0.4, so it needs to be dealt with somehow now.

File: src/module.c [hunk 4]
```diff
@@ -542,9 +542,14 @@ void jl_module_run_initializer(jl_module_t *m)
...
-        jl_printf(JL_STDERR, "WARNING: error initializing module %s:\n", m->name->name);
-        jl_static_show(JL_STDERR, jl_exception_in_transit);
-        jl_printf(JL_STDERR, "\n");
+        if (jl_initerror_type == NULL) {
+            jl_rethrow();
+        }
+        else {
+            jl_rethrow_other(jl_new_struct(jl_initerror_type, m->name,
+                                           jl_exception_in_transit));
+            jl_printf(JL_STDERR, "\n");
```
Comment (JeffBezanson) [inline src/module.c]: It's really weird to print something if the exception isn't being handled. If you tried to trap and hide all of these exceptions, you would still get empty lines scrolling in stderr.

File: src/toplevel.c [hunk 5]
```diff
@@ -585,8 +592,13 @@ jl_value_t *jl_parse_eval_all(const char *fname, size_t len)
             jl_rethrow();
         }
         else {
-            jl_rethrow_other(jl_new_struct(jl_loaderror_type, fn, ln,
-                                           jl_exception_in_transit));
+            if(jl_typeis(jl_exception_in_transit, jl_initerror_type)) {
```
Comment (JeffBezanson) [inline src/toplevel.c]: I don't think this makes sense. Whether an error causes loading a file to fail is orthogonal to the underlying cause of the error. For example if this is "x.jl":

```
using Foo  # throws InitError

Foo.f()
```

we won't get a LoadError for loading x.jl.

Comment (phobon) [inline src/toplevel.c]: You're right, I think I implicitly assumed this would only get used in one-file-one-module scenarios.

It seems the most appropriate thing would be to run the initializers only after all the code has been loaded, but I suppose that is quite a major change. Is a doubly wrapped exception what makes the most sense then?

Comment (stevengj) [inline src/toplevel.c]: Maybe just ditch `InitError` and use `LoadError` for `__init__` errors?  After all, a `LoadError` is what you would get from the same code pasted at the end of the module.

Comment (JeffBezanson) [inline src/toplevel.c]: Just double-wrapping the exception is probably better. That way the decision to have InitError is orthogonal to how everything else works. I'm not sure what you can do by catching InitErrors, but at least they provide more information for printing exceptions, which seems harmless enough for now.

**Comment by JeffBezanson** [comment]
@tkelman Very good point. All I can think of at the moment is to wrap those specific Base __init__ methods in try..catch to avoid the breakage.

It's also quite likely I was thinking of this issue when I put the try-catch in there originally. Before __init__ existed, we had the property that loading a module and never using it was unlikely to cause problems. With __init__, you are forced to run code in the module.

**Comment by phobon** [comment]
To be continued in #12742 ...
